src/basic/fs-util.c

```diff
@@ -1256,3 +1256,101 @@ int link_fd(int fd, int newdirfd, const char *newpath) {
 
         return RET_NERRNO(linkat(fd, "", newdirfd, newpath, AT_EMPTY_PATH));
 }
+
+int linkat_replace(int olddirfd, const char *oldpath, int newdirfd, const char *newpath) {
+        _cleanup_close_ int old_fd = -EBADF;
+        int r;
+
+        /* Like linkat() but replaces the target if needed. Is a NOP if source and target already share the
+         * same inode. */
+
+        if (olddirfd == AT_FDCWD && isempty(oldpath)) /* Refuse operating on the cwd (which is a dir, and dirs can't be hardlinked) */
+                return -EISDIR;
+
+        if (oldpath) {
+                if (STR_IN_SET(oldpath, ".", "..")) /* Refuse these definite directories early */
+                        return -EISDIR;
+
+                if (ENDSWITH_SET(oldpath, "/", "/.", "/..")) /* ditto */
+                        return -EISDIR;
+        }
+
+        if (isempty(newpath)) /* source path is optional, but the target path is not */
+                return -EINVAL;
+
+        if (STR_IN_SET(newpath, ".", ".."))
+                return -EISDIR;
+
+        if (ENDSWITH_SET(newpath, "/", "/.", "/..")) /* ditto */
+                return -EISDIR;
```

Comment:
It would be nicer to extract this into a helper in `path-util`?

Reply:
indeed